### PR TITLE
os/FileStore: fix to handle readdir error correctly

### PR DESCRIPTION
**Pull request by t-msn**
Currently filestore code does not handle readdir error.
As man readdir(3) says, we need to check errno after readdir
returns NULL to determine if error happens or not.

This patch fixes the all readdir() calls to check errono and
handle it appropriately:
 - FileStore.cc ... abort if EIO error happens
 - BtrfsFileStoreBAckend.cc/LFNindex.cc
                ... return error to upper layer

Without this fixes, primary PG could fail to correctly perform
backfill operation and could lead data loss propagation described
in #50558.

Fixes: https://tracker.ceph.com/issues/50558
Signed-off-by: Misono Tomohiro <misono.tomohiro@jp.fujitsu.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by t-msn** [comment]
I checked that upper layer of BtrfsBackend.cc does correctly handle EIO error (i.e abort if EIO)
but I have not yet confirmed it for LFNindex.cc. I think I need to check it as well.

File: src/os/filestore/LFNIndex.cc
Comment (tchaikov) [inline src/os/filestore/LFNIndex.cc:486]: nit, might want to move the definition of `r` closer to where it is used. i.e., to move to right before the `while()` loop.

**Review by tchaikov — APPROVED** [review]

File: src/os/filestore/LFNIndex.cc
Comment (t-msn) [reply src/os/filestore/LFNIndex.cc:486]: thx, I will update code

Comment (t-msn) [reply src/os/filestore/LFNIndex.cc]: pushed.

File: src/os/filestore/LFNIndex.cc
Comment (trociny) [inline src/os/filestore/LFNIndex.cc:438]: nit: no need to use `c_str`. Just `<< to_list_path` would be fine.

File: src/os/filestore/LFNIndex.cc
Comment (trociny) [inline src/os/filestore/LFNIndex.cc:439]: nit: `cpp_strerror(r)` would work too.

File: src/os/filestore/LFNIndex.cc
Comment (trociny) [inline src/os/filestore/LFNIndex.cc:499]: nit: `cpp_strerror(r)`

File: src/os/filestore/LFNIndex.cc
Comment (trociny) [inline src/os/filestore/LFNIndex.cc:498]: nit: `<< to_list_path <<`

File: src/os/filestore/FileStore.cc
Comment (trociny) [inline src/os/filestore/FileStore.cc:4996]: nit: I would prefer `cpp_strerror(r)`, but you can ignore (here and below).

**Review by trociny — APPROVED** [review]
Otherwise LGTM. Thanks!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-30_03:09:53-rados-wip-kefu-testing-2021-04-30-0020-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-30_10:06:45-rados-wip-kefu-testing-2021-04-30-1335-distro-basic-smithi/

**Review by jdurgin — APPROVED** [review]

**Review by neha-ojha — APPROVED** [review]

File: src/os/filestore/FileStore.cc
Comment (t-msn) [reply src/os/filestore/FileStore.cc:4996]: Hi, thanks for the review.

Actually I followed the style of other code in the file. So I'd like to keep this if it is ok.

File: src/os/filestore/LFNIndex.cc
Comment (t-msn) [reply src/os/filestore/LFNIndex.cc]: Fixed & pushed. Thanks.

File: src/os/filestore/LFNIndex.cc
Comment (t-msn) [reply src/os/filestore/LFNIndex.cc]: fixed this too. thanks.

**Comment by t-msn** [comment]
@tchaikov  Hello, thanks for the review & test.
As I am new to ceph community, could I ask some question about merge/backport process?
 - My understanding is currently this PR is under test. After finishing test, will this be merged? How long does it usually take to finish test?
 - After merged to master, could I need to do something for backport?, or will only backport team handle it? 
 
 If there is something I could help, please let me know.
 Thanks

**Comment by tchaikov** [comment]
> @tchaikov Hello, thanks for the review & test.

@t-msn, hello Misono.

> As I am new to ceph community, could I ask some question about merge/backport process?
> 

sure.

> * My understanding is currently this PR is under test. After finishing test, will this be merged? How long does it usually take to finish test?

the test takes couple days. see https://pulpito.ceph.com/kchai-2021-05-05_13:07:31-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/ for the first run of the batch which includes your PR. that test was finished 2 days ago. 

but sometimes, the test fails due to various reasons ranges from infra issues to known bugs. and sometimes, the test failures could be introduced by the PRs under tests. in general, we 

- analyze the test results, and
- try to minimize the impact of infra issues by rerunning the failed test.
- need to ensure that we have good coverage.

> * After merged to master, could I need to do something for backport?, or will only backport team handle it?
> 

you can always help with the backport. and yes, the stable release team handles it in general. but please note, anyone who's willing to contribute can be a team member =)

> If there is something I could help, please let me know.
> Thanks

**Comment by t-msn** [comment]
@tchaikov  Thank you very much for the information. 

> the test takes couple days. see https://pulpito.ceph.com/kchai-2021-05-05_13:07:31-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/ for the first run of the batch which includes your PR. that test was finished 2 days ago.
> 
> but sometimes, the test fails due to various reasons ranges from infra issues to known bugs. and sometimes, the test failures could be introduced by the PRs under tests. 

I have one more question. Can I search which test jobs include a certain commit at https://pulpito.ceph.com? 
I'm not sure which is the second bach test including this PR.

> you can always help with the backport. and yes, the stable release team handles it in general. but please note, anyone who's willing to contribute can be a team member =)

I checked https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst and it seems this commit wold be straightforward enough to be handled by scripts. But If manual backport is needed, I could do that.

Thanks,

**Comment by tchaikov** [comment]
> I have one more question. Can I search which test jobs include a certain commit at https://pulpito.ceph.com?
> I'm not sure which is the second bach test including this PR.

there is no straightforward way to search a test job by the PR under test. i just scheduled the second batch. see https://pulpito.ceph.com/kchai-2021-05-07_01:47:01-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/

**Comment by trociny** [comment]
@t-msn I was planning to take care of the backport when it was merged. Yes, as you said, in this case it will be a little effort, just a matter of running the script. But if you want to do this yourself by some reason (e.g. to become familiar with the process), just let me know and I leave it to you.

**Comment by t-msn** [comment]
@trociny Hi,
>  I was planning to take care of the backport when it was merged. Yes, as you said, in this case it will be a little effort, just a matter of running the script. 

Then, I'm happy that you will take care of the backport at this time. Thank you very much.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-05_13:07:31-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-07_01:47:01-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-07_12:44:46-rados-wip-kefu-testing2-2021-05-07-1516-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-08_03:28:06-rados-wip-kefu-testing2-2021-05-07-1516-distro-basic-smithi/

**Pull request merged**
